internal/provider/segment_resource.go

```diff
@@ -0,0 +1,159 @@
+package provider
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/resource"
+	"github.com/hashicorp/terraform-plugin-framework/resource/schema"
+	"github.com/hashicorp/terraform-provider-grafana-adaptive-metrics/internal/client"
+	"github.com/hashicorp/terraform-provider-grafana-adaptive-metrics/internal/model"
+)
+
+type segmentResource struct {
+	client *client.Client
+}
+
+var (
+	_ resource.Resource                = &segmentResource{}
+	_ resource.ResourceWithConfigure   = &segmentResource{}
+	_ resource.ResourceWithImportState = &segmentResource{}
+)
+
+func newSegmentResource() resource.Resource {
+	return &segmentResource{}
+}
+
+func (e *segmentResource) Configure(_ context.Context, req resource.ConfigureRequest, resp *resource.ConfigureResponse) {
+	if req.ProviderData == nil {
+		return
+	}
+
+	data, ok := req.ProviderData.(*resourceData)
+	if !ok {
+		resp.Diagnostics.AddError(
+			"Unexpected resource configure type",
+			fmt.Sprintf("Got %T. Please report this issue to the provider developers.", req.ProviderData),
+		)
+		return
+	}
+
+	e.client = data.client
+}
+
+func (e *segmentResource) Metadata(_ context.Context, req resource.MetadataRequest, resp *resource.MetadataResponse) {
+	resp.TypeName = fmt.Sprintf("%s_segment", req.ProviderTypeName)
```

Comment:
I'm not familiar with TF conventions, but wouldn't `grafana-adaptive-metrics-segment` look better than `grafana-adaptive-metrics_segment`? Or is it a convention to use `_` here?

Reply:
So there's only one real "rule" as far as I can tell: resource names are `<provider name>_<resource name>`. The underscore here is required, and the provider name is already set to `grafana-adaptive-metrics`. Changing it now is a breaking change, so we're stuck unless we want to do a deprecation + migration docs. In hindsight, it should have been `grafana_adaptive_metrics_segment` or even `grafanaam_segment` which would look like many other provider names.